safenode/src/transfers/error.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::protocol::types::{address::DbcAddress, fees};
+
+use sn_dbc::{Error as DbcError, SignedSpend, Token};
+
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeSet;
+use thiserror::Error;
+
+/// The type returned by the `sn_routing` message handling methods.
+pub(crate) type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Transfer errors.
+#[derive(Error, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum Error {
+    #[error("The transfer fee is missing.")]
+    MissingFee,
+    #[error("Invalid fee blinded amount.")]
+    InvalidFeeBlindedAmount,
+    #[error("Too low amount for the transfer fee: {paid}. Min required: {required}.")]
+    FeeTooLow { paid: Token, required: Token },
+    #[error(transparent)]
+    Fees(#[from] fees::Error),
+    #[error("Contacting close group of parent spends failed: {0}.")]
+    SpendParentCloseGroupIssue(String),
+    /// An error from the `sn_dbc` crate.
+    #[error("Dbc error: {0}")]
+    Dbcs(String),
+    /// Spend not found.
+    #[error("Spend not found: {0:?}")]
+    SpendNotFound(DbcAddress),
+    /// A double spend attempt was detected.
+    #[error("A double spend attempt was detected. Incoming and existing spend are not the same: {new:?}. Existing: {existing:?}")]
+    DoubleSpendAttempt {
+        /// New spend that we received.
+        new: Box<SignedSpend>,
+        /// Existing spend of same id that we already have.
+        existing: Box<SignedSpend>,
+    },
+    /// We were notified about a double spend attempt, but they were for different dbcs.
```

Comment:
Would it be worth adding something to the doc comments to explain how this error could occur? The name is kind of begging the question, but it doesn't seem obvious why something you thought was a double spend would actually turn out not to be.

Reply:
It's just a path handled. When there are two items and we can't have type safety that ensures the dbc ids are the same.
Then we must validate: are the ids the same?

If they are not, then that is a path that must be handled some way, even though we don't have a path in our code to reach that, since the msg could be sent with malice.

It can be seen where the error is raised.

Reply:
Yeah, I saw the code where the error occurred and it still wasn't clear to me why this would occur.

Reply:
Okay, well. I'll try to clarify that then.